libc-test/build.rs

```diff
@@ -1866,3 +1867,67 @@ fn test_dragonflybsd(target: &str) {
 
     cfg.generate("../src/lib.rs", "main.rs");
 }
+
+fn test_wasi(target: &str) {
+    assert!(target.contains("wasi"));
+
+    let mut cfg = ctest::TestGenerator::new();
+    cfg.define("_GNU_SOURCE", None);
+
+    headers! { cfg:
+        "errno.h",
+        "fcntl.h",
+        "limits.h",
+        "locale.h",
+        "malloc.h",
+        "stddef.h",
+        "stdint.h",
+        "stdio.h",
+        "stdlib.h",
+        "sys/stat.h",
+        "sys/types.h",
+        "time.h",
+        "unistd.h",
+        "wasi/core.h",
+        "wasi/libc.h",
+        "wchar.h",
+    }
+
+    cfg.type_name(move |ty, is_struct, is_union| match ty {
+        "FILE" => ty.to_string(),
+        t if is_union => format!("union {}", t),
+        t if t.starts_with("__wasi") && t.ends_with("_u") => format!("union {}", t),
+        t if t.starts_with("__wasi") && is_struct => format!("struct {}", t),
+        t if t.ends_with("_t") => t.to_string(),
+        t if is_struct => format!("struct {}", t),
+        t => t.to_string(),
+    });
+
+    // This is an opaque struct but we go through shenanigans to define values
+    // for it
+    cfg.skip_struct(move |ty| ty == "__clockid");
+
+    cfg.field_name(move |_struct, field| {
+        match field {
+			// deal with fields as rust keywords
+            "type_" => "type".to_string(),
+            s => s.to_string(),
+        }
+    });
+
+    cfg.skip_static(move |name| {
+        match name {
+            // wasi shenanigans for defining CLOCK_REALTIME and such
+            s if s.starts_with("__CLOCK") => true,
+            _ => false,
+        }
+    });
+
+    // Looks like LLD doesn't merge duplicate imports, so if the Rust
+    // code imports from a module and the C code also imports from a
+    // module we end up with two imports of function pointers which
+    // improt the same thing bug have different function pointers
```

Comment:
typo: "bug"/"but" 
typo: "improt"/"import"

So IIUC the issue here is that "libc" (wasi/the sysroot) is linked statically into both the Rust code (due to "-C target-feature=+crt-static"), and the C code, and that when these two are linked, two copies of the "libc" are present in the final binary. This is super-unfortunate. 

Have you tried whether `-C lto=fat` solves the issue ? That might be too brittle. Alternatively, maybe for this target it might make sense to not link the C code (or the Rust code) to the sysroot when compiling, so that when Rust and C are linked together, there is only one sysroot.

Reply:
Nah this is actually a different issue that's somewhat deeper in LLD and unrelated to linkage/LTO/etc. It has to do with an LLVM module that looks like:

```llvm
define void @foo() {
    call void @bar1()
    call void @bar2()
}

declare void @bar1() #1 
declare void @bar2() #2 

attributes #1 = { "wasm-import-module"="foo" }
attributes #1 = { "wasm-import-module"="foo" "wasm-import-name"="bar1" }
```

Both `bar1` and `bar2` reference the same function, but due to how LLD is implemented these show up as two different functions in the final binary, where the same name is imported twice (giving the same function two different function indices).

It's not really an issue, just a minor thing that will likely get fixed in LLD at some point.